c10/core/impl/CUDATraceTLS.cpp

```diff
@@ -0,0 +1,37 @@
+#include <c10/core/impl/CUDATraceTLS.h>
+
+#include <mutex>
+
+namespace c10 {
+namespace impl {
+
+static std::atomic<const PyInterpreter*> cudaTraceState;
+
+void CUDATraceTLS::set_trace(const PyInterpreter* trace) {
+  static std::once_flag flag;
+  std::call_once(
+    flag,
+    [&](){ cudaTraceState.store(trace); }
+  );
+  cudaTraceState.store(trace);
+}
+
+const PyInterpreter* CUDATraceTLS::get_trace() {
+  return cudaTraceState.load();
+}
```

Comment:
I wonder if defining this function inline in the header could give us some performance benefit? I'm saying it because this function is always invoked, even in the fast-path, even when the hooks are off and everything is a no-op. Hence this is kind of the hottest function. If we define it in the header we allow the compiler to inline it and thus avoid whatever cost might be associated to the function call (a handful of instructions for (re)storing the stack pointer, ...)? It could be interesting to try it out and benchmark using callgrind to see if there's any difference.

Reply:
I think defining it as an inline function in the header would be better for perf indeed.
Also, under the assumption that these hooks will only be used for non-perf critical cases, we can optimize even more for the case where they're not used.
An example where we already do that are the saved variable hooks: https://github.com/pytorch/pytorch/blob/8b4fee59120b70589b043daadfe7d263d116b373/aten/src/ATen/SavedTensorHooks.cpp#L11-L15
The short version is:
- We have a global boolean that is off by default
- The first time any hook is registered, flip this boolean to true.
- Checking if the hooks are enabled is now done in two steps: check if this boolean is enabled and then check the TLS.
- The boolean is never flipped back to false to remove the need for atomicity/sync when manipulating it.

This ensures that the case where perf matters (these hooks are not used), the only cost we pay is a very predictable branch on a single global boolean. This should be really negligible.
In the case where hooks are used, this branch will also be super predictable and so that shouldn't slow down the case where hooks are used either.

Reply:
I've benchmarked the performance with the function in the header file and with the `C10_ALWAYS_INLINE` macro, and saw an improvement with having it in the header but no further improvement with the macro.

Reply:
I have now also benchmarked the performance with the global variable approach, using a simple benchmark testing the fastpath allocating a tensor and counting the instructions. If my results are correct it would improve the performance by 0.007%, but there's still some concern about the safety of unsynchronized access to the global variable. Please let me know if you'd like me to implement this approach.

Benchmark: https://github.com/sypneiwski/pytorch/commit/15133e355851af2c791d497334ab5836838d173b
Results: P521007046